File: system_api.py
"""
system_api.py — the routes for AURA changing itself and staying in step.

Three groups, mounted as one router so a failure here can't take the brain
down with it:

    /api/upgrade/*   AURA proposing changes to its own source, and to any
                     project open in the Domain. Nothing is written until a
                     proposal is approved, and anything applied can be rolled
                     back.
    /api/system/*    resetting AURA, scope by scope.
    /api/sources/*   the GitHub link, document folders and pages AURA keeps
                     synced, and the repos it can pull down to work on.
"""
from __future__ import annotations

import os
from typing import Any

from fastapi import APIRouter, Request

router = APIRouter()

AURA_ROOT = os.path.dirname(os.path.abspath(__file__))


def _err(msg: str) -> dict[str, Any]:
    return {"ok": False, "error": msg}


# ════════════════════════════════════════════════════════════════════════════
# Upgrades — AURA writing code, its own included
# ════════════════════════════════════════════════════════════════════════════
@router.get("/api/upgrade/proposals")
async def upgrade_list(scope: str = "", root: str = "", limit: int = 50) -> dict[str, Any]:
    from core import patcher
    return {"ok": True, "proposals": patcher.list_proposals(scope, root, limit)}


@router.get("/api/upgrade/proposals/{pid}")
async def upgrade_get(pid: str) -> dict[str, Any]:
    from core import patcher
    p = patcher.get(pid)
    return {"ok": bool(p), "proposal": p}


@router.post("/api/upgrade/propose")
async def upgrade_propose(req: Request) -> dict[str, Any]:
    """Ask for a change. `scope: "self"` points at AURA's own source; anything
    else needs a `root`. Returns the proposal — it is NOT applied."""
    from core import patcher
    body = await req.json()
    request_text = str(body.get("request") or "").strip()
    if not request_text:
        return _err("tell me what to change")

    scope = str(body.get("scope") or "project")
    if scope == "self":
        root = AURA_ROOT
    else:
        root = str(body.get("root") or "").strip()
        if not root:
            return _err("which project?")
        if not os.path.isdir(root):
            return _err("that folder isn't there")

    hint = body.get("files") or body.get("hint") or []
    if isinstance(hint, str):
        hint = [hint]
    return patcher.propose(root, request_text, scope=scope,
                           hint=[str(h) for h in hint][:6],
                           label=str(body.get("label") or ""))


@router.post("/api/upgrade/proposals/{pid}/approve")
async def upgrade_approve(pid: str) -> dict[str, Any]:
    from core import patcher
    return patcher.approve(pid)


@router.post("/api/upgrade/proposals/{pid}/reject")
async def upgrade_reject(pid: str) -> dict[str, Any]:
    from core import patcher
    return patcher.reject(pid)


@router.post("/api/upgrade/proposals/{pid}/rollback")
async def upgrade_rollback(pid: str) -> dict[str, Any]:
    from core import patcher
    return patcher.rollback(pid)


@router.delete("/api/upgrade/proposals/{pid}")
async def upgrade_delete(pid: str) -> dict[str, Any]:
    from core import patcher
    return patcher.delete(pid)


@router.get("/api/upgrade/files")
async def upgrade_files(root: str = "", limit: int = 1200) -> dict[str, Any]:
    """The editable files under a tree — what the file picker lists."""
    from core import patcher
    target = root or AURA_ROOT
    if not os.path.isdir(target):
        return _err("that folder isn't there")
    return {"ok": True, "root": os.path.realpath(target),
            "files": patcher.file_tree(target, limit)}


@router.get("/api/upgrade/file")
async def upgrade_file(path: str, root: str = "") -> dict[str, Any]:
    from core import patcher
    target = root or AURA_ROOT
    body = patcher.read_file(target, path)
    if body is None:
        return _err("can't read that file")
    return {"ok": True, "path": path, "content": body}


# ════════════════════════════════════════════════════════════════════════════
# Reset
# ════════════════════════════════════════════════════════════════════════════
@router.get("/api/system/reset")
async def reset_scopes() -> dict[str, Any]:
    from core import system_reset
    return {"ok": True, "scopes": system_reset.scopes()}


@router.post("/api/system/reset")
async def reset_run(req: Request) -> dict[str, Any]:
    """Clear the named scopes. `confirm` must be the word RESET — a stray
    POST must not be able to wipe someone's memory."""
    from core import system_reset
    body = await req.json()
    if str(body.get("confirm") or "").strip().upper() != "RESET":
        return _err("type RESET to confirm")
    scopes = body.get("scopes") or []
    if not isinstance(scopes, list):
        return _err("nothing was selected")
    return system_reset.reset([str(s) for s in scopes])


# ════════════════════════════════════════════════════════════════════════════
# Sources
# ════════════════════════════════════════════════════════════════════════════
@router.get("/api/sources")
async def sources_list(room_id: int | None = None) -> dict[str, Any]:
    from core import sources
    return {"ok": True, "sources": sources.list_sources(room_id)}


@router.post("/api/sources")
async def sources_add(req: Request) -> dict[str, Any]:
    """Connect a GitHub link, a folder of documents, or a page. The first sync
    starts immediately, in the background."""
    from core import sources
    body = await req.json()
    raw = str(body.get("url") or body.get("path") or "").strip()
    if not raw:
        return _err("paste a link or a folder path")
    room = body.get("room_id")
    return sources.add(raw, label=str(body.get("label") or ""),
                       room_id=int(room) if room is not None else None,
                       auto_sync=bool(body.get("auto_sync", True)))


@router.patch("/api/sources/{source_id}")
async def sources_update(source_id: int, req: Request) -> dict[str, Any]:
    from core import sources
    body = await req.json()
    return sources.update(
        source_id,
        label=body.get("label"),
        room_id=body["room_id"] if "room_id" in body else "keep",
        auto_sync=body.get("auto_sync"),
    )


@router.delete("/api/sources/{source_id}")
async def sources_remove(source_id: int) -> dict[str, Any]:
    from core import sources
    return sources.remove(source_id)


@router.post("/api/sources/{source_id}/sync")
async def sources_sync(source_id: int) -> dict[str, Any]:
    from core import sources
    return sources.sync(source_id)


@router.post("/api/sources/sync")
async def sources_sync_all() -> dict[str, Any]:
    """What the app calls on open."""
    from core import sources
    sources.sync_all()
    return {"ok": True, "started": True}


@router.get("/api/sources/repos")
async def sources_repos() -> dict[str, Any]:
    from core import sources
    return {"ok": True, "repos": sources.repos()}


@router.post("/api/sources/{source_id}/clone")
async def sources_clone(source_id: int, req: Request) -> dict[str, Any]:
    """Bring a repo to disk so it can be opened, prompted and edited."""
    from core import sources
    body = await req.json()
    full_name = str(body.get("full_name") or "").strip()
    if not full_name or "/" not in full_name:
        return _err("which repo?")
    return sources.clone(source_id, full_name)

File: frontend/src/systemApi.ts
// ============================================================================
// systemApi — the client for AURA changing itself and staying in step.
//
// Three groups, matching system_api.py: upgrades (AURA writing code, its own
// included), reset, and connected sources. Kept out of api.ts because none of
// it is about a conversation — it is about the machine underneath one.
// ============================================================================
import { BASE } from "./api";

async function j<T>(path: string, init?: RequestInit): Promise<T> {
  const res = await fetch(BASE + path, {
    headers: { "Content-Type": "application/json" },
    ...init,
  });
  return res.json() as Promise<T>;
}

const post = (body?: unknown): RequestInit =>
  ({ method: "POST", ...(body === undefined ? {} : { body: JSON.stringify(body) }) });

// ---- upgrades --------------------------------------------------------------

/** One file a proposal would rewrite. `before`/`after` are whole files. */
export interface FileChange {
  path: string;
  new: boolean;
  before: string;
  after: string;
  diff: string;
  added: number;
  removed: number;
}

export type ProposalStatus = "pending" | "applied" | "rejected" | "rolled_back";

/** The summary form, as the list endpoint returns it. */
export interface ProposalRow {
  id: string;
  /** "self" = AURA's own source. Anything else is a project in the Domain. */
  scope: string;
  root: string;
  label: string;
  request: string;
  why: string;
  status: ProposalStatus;
  created_at: string;
  applied_at: string;
  files: string[];
  added: number;
  removed: number;
}

/** The full form, with every file's before/after. */
export interface Proposal extends Omit<ProposalRow, "files"> {
  changes: FileChange[];
}

export interface ProposeResult extends Partial<Proposal> {
  ok: boolean;
  error?: string;
}

export const upgrades = {
  list: (scope = "", root = "") =>
    j<{ ok: boolean; proposals: ProposalRow[] }>(
      `/api/upgrade/proposals?scope=${encodeURIComponent(scope)}&root=${encodeURIComponent(root)}`,
    ).then((r) => r.proposals ?? []),

  get: (id: string) =>
    j<{ ok: boolean; proposal: Proposal | null }>(`/api/upgrade/proposals/${id}`)
      .then((r) => r.proposal),

  /** Ask for a change. Nothing is written — this returns a proposal to review. */
  propose: (body: { request: string; scope?: "self" | "project"; root?: string; files?: string[]; label?: string }) =>
    j<ProposeResult>("/api/upgrade/propose", post(body)),

  /** `restart` comes back true when AURA rewrote its own source. */
  approve: (id: string) =>
    j<{ ok: boolean; error?: string; applied?: string[]; restart?: boolean }>(
      `/api/upgrade/proposals/${id}/approve`, post()),

  reject: (id: string) =>
    j<{ ok: boolean; error?: string }>(`/api/upgrade/proposals/${id}/reject`, post()),

  rollback: (id: string) =>
    j<{ ok: boolean; error?: string; restored?: string[]; failed?: string[]; restart?: boolean }>(
      `/api/upgrade/proposals/${id}/rollback`, post()),

  remove: (id: string) =>
    j<{ ok: boolean; error?: string }>(`/api/upgrade/proposals/${id}`, { method: "DELETE" }),

  files: (root = "") =>
    j<{ ok: boolean; root: string; files: string[] }>(
      `/api/upgrade/files?root=${encodeURIComponent(root)}`),

  file: (path: string, root = "") =>
    j<{ ok: boolean; content?: string; error?: string }>(
      `/api/upgrade/file?path=${encodeURIComponent(path)}&root=${encodeURIComponent(root)}`),
};

// ---- reset -----------------------------------------------------------------

export interface ResetScope {
  id: string;
  label: string;
  desc: string;
}

export const system = {
  resetScopes: () =>
    j<{ ok: boolean; scopes: ResetScope[] }>("/api/system/reset").then((r) => r.scopes ?? []),

  /** `confirm` must be the literal word RESET. */
  reset: (scopes: string[], confirm: string) =>
    j<{ ok: boolean; error?: string; cleared?: string[]; labels?: string[]; backup?: string }>(
      "/api/system/reset", { method: "POST", body: JSON.stringify({ scopes, confirm }) }),
};

// ---- sources ---------------------------------------------------------------

export type SourceKind = "github" | "docs" | "link";
export type SourceStatus = "idle" | "syncing" | "ok" | "error";

export interface RepoRef {
  full_name: string;
  name: string;
  description: string;
  language: string;
  default_branch: string;
  updated_at: string;
  pushed_at: string;
  html_url: string;
  private: boolean;
  stars: number;
  /** Set once the repo has been cloned — the folder it lives in. */
  local?: string;
  pulled?: string;
  source_id?: number;
}

export interface Source {
  id: number;
  kind: SourceKind;
  url: string;
  label: string;
  room_id: number | null;
  auto_sync: boolean;
  status: SourceStatus;
  /** Human sentence about the last sync — "8 repos, 2 cloned locally". */
  detail: string;
  meta: {
    owner?: string;
    repo?: string;
    repos?: RepoRef[];
    files?: Record<string, { name: string; item: number; stamp: string }>;
    item?: number;
  };
  last_sync: string;
  created_at: string;
}

export const sources = {
  list: (roomId?: number | null) =>
    j<{ ok: boolean; sources: Source[] }>(
      "/api/sources" + (roomId != null ? `?room_id=${roomId}` : ""),
    ).then((r) => r.sources ?? []),

  add: (url: string, opts: { label?: string; room_id?: number | null } = {}) =>
    j<{ ok: boolean; error?: string; source?: Source }>("/api/sources", post({ url, ...opts })),

  update: (id: number, patch: { label?: string; room_id?: number | null; auto_sync?: boolean }) =>
    j<{ ok: boolean; source?: Source }>(`/api/sources/${id}`, {
      method: "PATCH", body: JSON.stringify(patch),
    }),

  remove: (id: number) => j<{ ok: boolean }>(`/api/sources/${id}`, { method: "DELETE" }),

  sync: (id: number) => j<{ ok: boolean; source?: Source }>(`/api/sources/${id}/sync`, post()),

  /** Fire-and-forget: what the app calls on open. */
  syncAll: () => j<{ ok: boolean }>("/api/sources/sync", post()),

  repos: () => j<{ ok: boolean; repos: RepoRef[] }>("/api/sources/repos").then((r) => r.repos ?? []),

  /** Bring a repo to disk so it can be opened, prompted and edited. */
  clone: (id: number, fullName: string) =>
    j<{ ok: boolean; error?: string; path?: string; already?: boolean }>(
      `/api/sources/${id}/clone`, post({ full_name: fullName })),
};
